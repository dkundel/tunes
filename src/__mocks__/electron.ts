import EventEmitter = require('events');

export class MockIpcRenderer extends EventEmitter {}

export const _ipcRendererEmitter = new MockIpcRenderer();

export const ipcRenderer = {
  send: jest.fn(),
  on: jest.fn((evt: string, callback: Function) =>
    _ipcRendererEmitter.on(evt, callback)
  )
};

export const shell = {
  openExternal: jest.fn()
};
