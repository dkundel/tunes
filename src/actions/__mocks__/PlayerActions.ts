export const togglePlayer = jest.fn(() => ({ type: 'TOGGLE_PLAYER' }));
export const nextSong = jest.fn(() => ({ type: 'NEXT_SONG' }));
export const prevSong = jest.fn(() => ({ type: 'PREV_SONG' }));
export const loadPlaylist = jest.fn((playlist: string) => ({
  type: 'LOAD_PLAYLIST',
  playlist
}));
export const addSong = jest.fn((song: string) => ({ type: 'ADD_SONG', song }));
