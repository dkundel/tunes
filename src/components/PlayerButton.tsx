import * as React from 'react';
import styled from 'styled-components';

export interface PlayerButtonProps {
  onClick?: Function;
  materialIcon: string;
}

const StyledButton = styled.button`
  width: 40px;
  height: 51px;
  border: 0;
  outline: 0;
  cursor: pointer;
  background: var(--colorPrimary);
  color: var(--colorSecondary);
  &:hover {
    background: var(--colorPrimaryHover);
  }
`;

export default class PlayerButton extends React.Component<
  PlayerButtonProps,
  undefined
> {
  public render() {
    return (
      <StyledButton onClick={() => this.props.onClick && this.props.onClick()}>
        <i className="material-icons">
          {this.props.materialIcon || 'help_outline'}
        </i>
      </StyledButton>
    );
  }
}
