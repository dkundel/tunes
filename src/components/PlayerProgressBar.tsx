import * as React from 'react';

export interface PlayerProgressBarProps {
  percentage: number;
  title: string;
}

export default class PlayerProgressBar extends React.Component<
  PlayerProgressBarProps,
  undefined
> {
  public render() {
    return (
      <div className="player-progress" ref="progressContainer">
        <p className="player-title">{this.props.title}</p>
        <div
          className="player-progress-bar"
          ref="progressBar"
          style={{ '--percentage': this.props.percentage + '%' }}
        />
        <div
          className="player-title-overlay"
          ref="titleOverlay"
          style={{ '--percentage': this.props.percentage + '%' }}
        >
          <p className="player-title-overlay-content">
            {this.props.title}
          </p>
        </div>
      </div>
    );
  }
}
