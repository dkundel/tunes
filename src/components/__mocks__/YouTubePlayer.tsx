import * as React from 'react';

export default class MockYouTubePlayer extends React.Component<any, any> {
  render() {
    return <div id="YouTubePlayer" />;
  }
}
