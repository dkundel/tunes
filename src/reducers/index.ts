import { combineReducers, Action } from 'redux';

import player from './player';

const rootReducer = combineReducers<Action>({
  player
});

export * from './state-types';

export default rootReducer;
