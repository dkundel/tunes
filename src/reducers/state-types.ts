export type PlayerStatus =
  | 'stopped'
  | 'paused'
  | 'playing'
  | 'loading'
  | 'resuming'
  | undefined;

export interface SongInfo {
  title: string;
  totalTime: number;
  percentage: number;
  currentTime: number;
  videoId: string;
}

export interface PlayerState {
  currentSong: SongInfo | undefined;
  currentPlayList: string | undefined;
  currentPlayerStatus: PlayerStatus;
  songQueue: string[];
  queueIndex: number | undefined;
  nextPlayerStatus: PlayerStatus;
}

export interface GlobalState {
  player: PlayerState;
}
