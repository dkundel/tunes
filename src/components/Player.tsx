import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { ipcRenderer } from 'electron';
import * as moment from 'moment';
import styled from 'styled-components';

import {
  togglePlayer,
  nextSong,
  prevSong,
  loadPlaylist,
  addSong
} from '../actions/PlayerActions';
import { GlobalState, SongInfo, PlayerStatus } from '../reducers';
import { getCurrentSongInfo, getPlayerStatus } from '../reducers/player';

import BlurryYouTubeBackground from './BlurryYouTubeBackground';
import PlayerButton from './PlayerButton';
import PlayerProgressBar from './PlayerProgressBar';
import YouTubePlayer from './YouTubePlayer';

export interface PlayerProps {
  currentSongInfo: SongInfo | undefined;
  playerStatus: PlayerStatus;
  togglePlayer: () => any;
  nextSong: () => any;
  prevSong: () => any;
  loadPlaylist: (playlist: string) => any;
  addSong: (playlist: string) => any;
}

export class Player extends React.Component<PlayerProps, undefined> {
  constructor(props: PlayerProps) {
    super(props);
    this.togglePlayPause = this.togglePlayPause.bind(this);
    this.nextSong = this.nextSong.bind(this);
    this.prevSong = this.prevSong.bind(this);
  }

  public componentDidMount() {
    ipcRenderer.on('player:toggle', () => {
      this.togglePlayPause();
    });
    ipcRenderer.on('player:prev', () => {
      this.prevSong();
    });
    ipcRenderer.on('player:next', () => {
      this.nextSong();
    });
    ipcRenderer.on('player:load:playlist', (_: any, playlist: string) => {
      this.props.loadPlaylist(playlist);
    });
    ipcRenderer.on('player:load:video', (_: any, song: string) => {
      this.props.addSong(song);
    });
  }

  public togglePlayPause() {
    this.props.togglePlayer();
  }

  public nextSong() {
    this.props.nextSong();
  }

  public prevSong() {
    this.props.prevSong();
  }

  private getTimeString() {
    let timeInSeconds = this.props.currentSongInfo &&
      this.props.currentSongInfo.currentTime
      ? this.props.currentSongInfo.currentTime
      : 0;

    const format = timeInSeconds >= 60 * 60 ? 'H:mm:ss' : 'mm:ss';
    return moment({
      hours: 0,
      minutes: 0,
      seconds: 0
    })
      .seconds(timeInSeconds)
      .format(format);
  }

  private getPercentage(): number {
    if (!this.props.currentSongInfo) {
      return 0;
    }
    return this.props.currentSongInfo.percentage;
  }

  private getTitle(): string {
    if (!this.props.currentSongInfo) {
      return 'No video loaded';
    }

    return this.props.currentSongInfo.title;
  }

  private getPlayButtonIcon(): string {
    const { playerStatus } = this.props;
    if (playerStatus === 'playing') {
      return 'pause';
    }

    if (playerStatus === 'paused' || playerStatus === 'stopped') {
      return 'play_arrow';
    }

    if (playerStatus === 'loading') {
      return 'loop';
    }

    return 'play_arrow';
  }

  private getVideoId() {
    if (!this.props.currentSongInfo) {
      return '';
    }
    return this.props.currentSongInfo.videoId;
  }

  public render() {
    return (
      <PlayerContainer>
        <PlayerBar>
          <PlayerButtonGroup>
            <PlayerButton
              onClick={() => this.togglePlayPause()}
              materialIcon={this.getPlayButtonIcon()}
            />
            <PlayerButton
              onClick={() => this.prevSong()}
              materialIcon="skip_previous"
            />
            <PlayerButton
              onClick={() => this.nextSong()}
              materialIcon="skip_next"
            />
          </PlayerButtonGroup>
          <PlayerProgressBar
            percentage={this.getPercentage()}
            title={this.getTitle()}
          />
          <PlayerMetaBar>
            <PlayerTimer className="timer">
              {this.getTimeString()}
            </PlayerTimer>
          </PlayerMetaBar>
        </PlayerBar>
        <BlurryYouTubeBackground videoId={this.getVideoId()} />
        <YouTubePlayer />
      </PlayerContainer>
    );
  }
}

const PlayerContainer = styled.div`
  width: 100%;
`;

const PlayerBar = styled.div`
  display: flex;
  color: var(--colorPrimary);
  background: rgba(255, 255, 255, 0.3);
`;

const PlayerButtonGroup = styled.div``;

const PlayerMetaBar = styled.div`
  background: var(--colorPrimary);
  color: var(--colorSecondary);
`;

const PlayerTimer = styled.p`
  margin-left: 15px;
  margin-right: 15px;
`;

export const mapStateToProps = (state: GlobalState) => ({
  currentSongInfo: getCurrentSongInfo(state),
  playerStatus: getPlayerStatus(state)
});

export const mapDispatchToProps = (dispatch: Dispatch<undefined>) => ({
  togglePlayer: () => dispatch(togglePlayer()),
  nextSong: () => dispatch(nextSong()),
  prevSong: () => dispatch(prevSong()),
  loadPlaylist: (playlist: string) => dispatch(loadPlaylist(playlist)),
  addSong: (song: string) => dispatch(addSong(song))
});

export default connect(mapStateToProps, mapDispatchToProps)(Player);
