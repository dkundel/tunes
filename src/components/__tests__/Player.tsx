jest.mock('../YouTubePlayer');
jest.mock('../../actions/PlayerActions');
jest.mock('electron');

import * as React from 'react';
import { ipcRenderer } from 'electron';
import { shallow } from 'enzyme';
import * as renderer from 'react-test-renderer';

import { Player, PlayerProps, mapDispatchToProps } from '../Player';
import { PlayerProgressBarProps } from '../PlayerProgressBar';
import { PlayerButtonProps } from '../PlayerButton';
import { SongInfo } from '../../reducers/state-types';
import {
  addSong,
  nextSong,
  prevSong,
  loadPlaylist,
  togglePlayer
} from '../../actions/PlayerActions';

import { _ipcRendererEmitter } from '../../__mocks__/electron';

describe('test Player component', () => {
  let props: PlayerProps;

  beforeEach(() => {
    props = {
      currentSongInfo: undefined,
      playerStatus: undefined,
      togglePlayer: jest.fn(),
      nextSong: jest.fn(),
      prevSong: jest.fn(),
      loadPlaylist: jest.fn(),
      addSong: jest.fn()
    };

    (ipcRenderer.on as jest.Mock<any>).mockClear();
  });

  test('handles default', () => {
    const component = renderer.create(<Player {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(ipcRenderer.on).toHaveBeenCalledTimes(5);
  });

  test('handles click events', () => {
    const component = shallow(<Player {...props} />);
    const buttons = component.find('PlayerButton');
    expect(buttons).toHaveLength(3);
    const playButton = buttons.at(0);
    const prevButton = buttons.at(1);
    const nextButton = buttons.at(2);
    playButton.simulate('click');
    expect(props.togglePlayer).toHaveBeenCalledTimes(1);
    prevButton.simulate('click');
    expect(props.prevSong).toHaveBeenCalledTimes(1);
    nextButton.simulate('click');
    expect(props.nextSong).toHaveBeenCalledTimes(1);
  });

  test('display song info correct [snapshot]', () => {
    const songInfo: SongInfo = {
      title: 'My Unit Test',
      currentTime: 0,
      percentage: 0,
      totalTime: 10,
      videoId: 'someid'
    };
    props.currentSongInfo = songInfo;
    const component = renderer.create(<Player {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('display song info correct [components]', () => {
    const songInfo: SongInfo = {
      title: 'My Unit Test',
      currentTime: 3,
      percentage: 33.3,
      totalTime: 10,
      videoId: 'someid'
    };
    props.currentSongInfo = songInfo;
    const component = shallow(<Player {...props} />);

    // has the right progress and title
    const progressBar = component.find('PlayerProgressBar');
    const progressBarProps = progressBar.props() as PlayerProgressBarProps;
    expect(progressBarProps.percentage).toBe(33.3);
    expect(progressBarProps.title).toBe('My Unit Test');

    // has the right time
    const playerTimer = component.find('.timer').render();
    expect(playerTimer.text()).toBe('00:03');
  });

  test('display song info correct for longer time [components]', () => {
    const songInfo: SongInfo = {
      title: 'My Unit Test',
      currentTime: 3603,
      percentage: 33.3,
      totalTime: 10,
      videoId: 'someid'
    };
    props.currentSongInfo = songInfo;
    const component = shallow(<Player {...props} />);

    // has the right time
    const playerTimer = component.find('.timer').render();
    expect(playerTimer.text()).toBe('1:00:03');
  });

  test('changes the button label correctly for different player states', () => {
    const songInfo: SongInfo = {
      title: 'My Unit Test',
      currentTime: 3,
      percentage: 33.3,
      totalTime: 10,
      videoId: 'someid'
    };
    props.currentSongInfo = songInfo;
    props.playerStatus = 'stopped';
    const component = shallow(<Player {...props} />);

    // stopped player
    let playButton = component.find('PlayerButton').at(0);
    expect((playButton.props() as PlayerButtonProps).materialIcon).toBe(
      'play_arrow'
    );

    // playing player
    component.setProps({
      ...props,
      playerStatus: 'playing'
    });
    playButton = component.find('PlayerButton').at(0);
    expect((playButton.props() as PlayerButtonProps).materialIcon).toBe(
      'pause'
    );

    // loading player
    component.setProps({
      ...props,
      playerStatus: 'loading'
    });
    playButton = component.find('PlayerButton').at(0);
    expect((playButton.props() as PlayerButtonProps).materialIcon).toBe('loop');

    // paused player
    component.setProps({
      ...props,
      playerStatus: 'paused'
    });
    playButton = component.find('PlayerButton').at(0);
    expect((playButton.props() as PlayerButtonProps).materialIcon).toBe(
      'play_arrow'
    );
  });

  test('calls the right functions based on events', () => {
    const originalToggle = Player.prototype.togglePlayPause;
    Player.prototype.togglePlayPause = jest
      .fn()
      .mockImplementation(originalToggle);
    const originalPrev = Player.prototype.prevSong;
    Player.prototype.prevSong = jest.fn().mockImplementation(originalPrev);
    const originalNext = Player.prototype.nextSong;
    Player.prototype.nextSong = jest.fn().mockImplementation(originalNext);
    _ipcRendererEmitter.removeAllListeners();

    renderer.create(<Player {...props} />);

    _ipcRendererEmitter.emit('player:toggle');
    expect(Player.prototype.togglePlayPause).toHaveBeenCalledTimes(1);

    _ipcRendererEmitter.emit('player:prev');
    expect(Player.prototype.prevSong).toHaveBeenCalledTimes(1);

    _ipcRendererEmitter.emit('player:next');
    expect(Player.prototype.nextSong).toHaveBeenCalledTimes(1);

    _ipcRendererEmitter.emit('player:load:playlist');
    expect(props.loadPlaylist).toHaveBeenCalledTimes(1);

    _ipcRendererEmitter.emit('player:load:video');
    expect(props.addSong).toHaveBeenCalledTimes(1);

    Player.prototype.togglePlayPause = originalToggle;
    Player.prototype.prevSong = originalPrev;
    Player.prototype.nextSong = originalNext;
  });
});

describe('test mapDispatchToProps', () => {
  const dispatch = jest.fn(() => true);
  const map = mapDispatchToProps(dispatch);

  beforeEach(() => {
    dispatch.mockClear();
    (togglePlayer as jest.Mock<any>).mockClear();
    (nextSong as jest.Mock<any>).mockClear();
    (prevSong as jest.Mock<any>).mockClear();
    (loadPlaylist as jest.Mock<any>).mockClear();
    (addSong as jest.Mock<any>).mockClear();
  });

  test('dispatches togglePlayer correctly', () => {
    map.togglePlayer();
    expect(dispatch).toHaveBeenCalledWith({ type: 'TOGGLE_PLAYER' });
    expect(togglePlayer).toHaveBeenCalledTimes(1);
  });

  test('dispatches nextSong correctly', () => {
    map.nextSong();
    expect(dispatch).toHaveBeenCalledWith({ type: 'NEXT_SONG' });
    expect(nextSong).toHaveBeenCalledTimes(1);
  });

  test('dispatches prevSong correctly', () => {
    map.prevSong();
    expect(dispatch).toHaveBeenCalledWith({ type: 'PREV_SONG' });
    expect(prevSong).toHaveBeenCalledTimes(1);
  });

  test('dispatches loadPlaylist correctly', () => {
    map.loadPlaylist('playlistid');
    expect(dispatch).toHaveBeenCalledWith({
      type: 'LOAD_PLAYLIST',
      playlist: 'playlistid'
    });
    expect(loadPlaylist).toHaveBeenCalledWith('playlistid');
  });

  test('dispatches addSong correctly', () => {
    map.addSong('songid');
    expect(dispatch).toHaveBeenCalledWith({
      type: 'ADD_SONG',
      song: 'songid'
    });
    expect(addSong).toHaveBeenCalledWith('songid');
  });
});
