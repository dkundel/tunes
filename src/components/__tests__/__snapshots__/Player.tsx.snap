// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Player component display song info correct [snapshot] 1`] = `
<div
  className="sc-gzVnrw lfdfns"
>
  <div
    className="sc-htoDjs hOPoSy"
  >
    <div
      className="sc-dnqmqq bDsesi"
    >
      <button
        className="sc-bwzfXH dzjsPu"
        onClick={[Function]}
      >
        <i
          className="material-icons"
        >
          play_arrow
        </i>
      </button>
      <button
        className="sc-bwzfXH dzjsPu"
        onClick={[Function]}
      >
        <i
          className="material-icons"
        >
          skip_previous
        </i>
      </button>
      <button
        className="sc-bwzfXH dzjsPu"
        onClick={[Function]}
      >
        <i
          className="material-icons"
        >
          skip_next
        </i>
      </button>
    </div>
    <div
      className="sc-htpNat dgMhXc"
    >
      <p
        className="sc-bxivhb iIkqZF"
      >
        My Unit Test
      </p>
      <div
        className="sc-EHOje eCGwaZ"
        style={
          Object {
            "--percentage": "0%",
          }
        }
      />
      <div
        className="sc-bZQynM hMKrxa"
        style={
          Object {
            "--percentage": "0%",
          }
        }
      >
        <p
          className="sc-ifAKCX jmWDig"
        >
          My Unit Test
        </p>
      </div>
    </div>
    <div
      className="sc-iwsKbI ktdSOd"
    >
      <p
        className="timer sc-gZMcBi kegliO"
      >
        00:00
      </p>
    </div>
  </div>
  <img
    className="player-background blur-background sc-bdVaJa boBCaA"
    src="https://img.youtube.com/vi/someid/default.jpg"
  />
  <div
    id="YouTubePlayer"
  />
</div>
`;

exports[`test Player component handles default 1`] = `
<div
  className="sc-gzVnrw lfdfns"
>
  <div
    className="sc-htoDjs hOPoSy"
  >
    <div
      className="sc-dnqmqq bDsesi"
    >
      <button
        className="sc-bwzfXH dzjsPu"
        onClick={[Function]}
      >
        <i
          className="material-icons"
        >
          play_arrow
        </i>
      </button>
      <button
        className="sc-bwzfXH dzjsPu"
        onClick={[Function]}
      >
        <i
          className="material-icons"
        >
          skip_previous
        </i>
      </button>
      <button
        className="sc-bwzfXH dzjsPu"
        onClick={[Function]}
      >
        <i
          className="material-icons"
        >
          skip_next
        </i>
      </button>
    </div>
    <div
      className="sc-htpNat dgMhXc"
    >
      <p
        className="sc-bxivhb iIkqZF"
      >
        No video loaded
      </p>
      <div
        className="sc-EHOje eCGwaZ"
        style={
          Object {
            "--percentage": "0%",
          }
        }
      />
      <div
        className="sc-bZQynM hMKrxa"
        style={
          Object {
            "--percentage": "0%",
          }
        }
      >
        <p
          className="sc-ifAKCX jmWDig"
        >
          No video loaded
        </p>
      </div>
    </div>
    <div
      className="sc-iwsKbI ktdSOd"
    >
      <p
        className="timer sc-gZMcBi kegliO"
      >
        00:00
      </p>
    </div>
  </div>
  <img
    className="player-background blur-background sc-bdVaJa boBCaA"
  />
  <div
    id="YouTubePlayer"
  />
</div>
`;
